drivers: adc: adc_nrfx_saadc: Patch neg single ended readings

bjarki-andreasen opened the pull request:
The NRF SAADC produces negative values in single ended mode if the positive input is below 0V (ground). This behavior does not match the ADC device driver API, which states that in single ended mode, the readings must be positive [0 .. 2^resolution - 1].

This commit extends the adc_nrfx_saadc device driver to track which channels in a sequence are configured to be single ended, to then adjust negative readings for these channels to be 0.

### Tests
Build adc_sequence sample with a modified board overlay which has one single ended and one differential channel configured.  The differential channel has the negative input tied to GND. The sample has also been modified to print signed values.

The positive inputs are then varied from -100mV to +100mV. The single ended channel should only show positive values, the differential channel should show negative and positive values. 

The test includes a nRF52840 for which all resolutions should be valid, and a nRF54L15 for which attempting to use 8-bit resolution on a single ended channel should fail to start the ADC read.

Fixes: #71119

#### nRF52840 (reference 600mV)
* 8 bit
  * differential
    * 100mV  = 21
    * -100mV = -22
  * single ended
    * 100mV = 65
    * -100mV = 0
* 10 bit
  * differential
    * 100mV  = 85
    * -100mV = -88
  * single ended
    * 100mV = 270
    * -100mV = 0
* 12 bit
  * differential
    * +100mV  = 340
    * -100mV = -340
  * single ended
    * +100mV = 1040
    * -100mV = 0
* 14 bit
  * differential
    * +100mV  = 1378
    * -100mV = -1368
  * single ended
    * +100mV = 4448
    * -100mV = 0

#### nRF54L15 (reference 900mV)
* 8 bit
  * differential
    * 100mV  = 15
    * -100mV = -15
    * 1000mV = 127
    * -1000mV = -128
  * single ended (fails to start as expected since sample size and res are both 8-bit)
    * 100mV = NA
    * -100mV = NA
* 10 bit
  * differential
    * 100mV  = 60
    * -100mV = -60
  * single ended
    * 100mV = 108
    * -100mV = 0
* 12 bit
  * differential
    * +100mV  = 228
    * -100mV = -230
  * single ended
    * +100mV = 420
    * -100mV = 0
* 14 bit
  * differential
    * +100mV  = 944
    * -100mV = -920
  * single ended
    * +100mV = 2128
    * -100mV = 0

bjarki-andreasen-nordic committed:
drivers: adc: adc_nrfx_saadc: Patch neg single ended readings

The NRF SAADC produces negative values in single ended mode if the
positive input is below 0V (ground). This behavior does not match
the ADC device driver API, which states that in single ended mode,
the readings must be positive [0 .. 2^resolution - 1].

This commit extends the adc_nrfx_saadc device driver to track which
channels in a sequence are configured to be single ended, to then
corrects negative readings for these channels to 0.

This patch only works if the ADC resolution is lower than the sample
bit size. This is the case for 8, 10, 12 and 14 bit resolutions for
the nRF 52, 53 and 91 series which store readings in a int16_t.

The nRF 54H and 54L series store 8-bit resolution readings in a 8-bit
sample size. A check has been added to start_read() to prevent single
ended mode readings if the resolution is 8-bit for these platforms.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
doc: release notes: add adc_nrfx_saadc patch note.

Add note describing patch which fixes issue which allowed negative
readings in single-ended mode.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -526,6 +583,13 @@ static int start_read(const struct device *dev,
 	adc_context_start_read(&m_data.ctx, sequence);
 
 	error = adc_context_wait_for_completion(&m_data.ctx);
+	if (error) {
+		return error;
+	}
+
+	if (selected_channels & m_data.single_ended_channels) {
+		correct_single_ended(sequence, active_channels);
```
This will not work for `adc_read_async()` because then `adc_context_wait_for_completion()` will most likely return before the samplings are done. Another problem will be with sequences that use a callback - at the time the callback is called, the values will not be corrected yet.
I think the correction should be done in `saadc_irq_handler()` instead.

bjarki-andreasen replied:
Right :) Moved correction to irq handler.

zagor commented:
I think you should also remove the workaround in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/sensor/voltage_divider/voltage.c#L62 in this patch.

bjarki-andreasen-nordic committed:
drivers: sensor: voltage_divider: revert patch

A patch was added to the voltage divider to handle an erroneous
negative voltage reading for the nrfx_saadc which could return
negative voltages even though the ADC mode was single ended.

This has now been patched in the ADC driver.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
reverted patch in `voltage_divider` sensor :)

bjarki-andreasen commented:
ping @anangl @masz-nordic :)

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -432,11 +445,42 @@ static int check_buffer_size(const struct adc_sequence *sequence,
+{
+	int8_t channel_id = -1;
+	uint32_t selected_channels = sequence->channels;
+	int16_t *sample_it = sequence->buffer;
+	int16_t *sample_it_end = sample_it + (active_channels * sum_samplings(sequence));
```
This is wrong. Only samples from one sampling should be corrected here (that one for which `saadc_irq_handler()` was called). I think you could move the call to `correct_single_ended()` to before that `#if defined(ADC_BUFFER_IN_RAM)` and then here you could get the buffer pointer with just `nrf_saadc_buffer_pointer_get(NRF_SAADC)`.

bjarki-andreasen replied:
I see, I will update it and test it on the 54H20 which seems to be the only one setting `ADC_BUFFER_IN_RAM` :)

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -412,16 +419,22 @@ static int set_oversampling(const struct adc_sequence *sequence,
-		needed_buffer_size *= (1 + sequence->options->extra_samplings);
-	}
+	needed_buffer_size = samples_to_bytes(sequence, active_channels)
+			   * sum_samplings(sequence);
```
This change is unneeded.

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -82,11 +82,12 @@ struct driver_data {
 	struct adc_context ctx;
 
 	uint8_t positive_inputs[SAADC_CH_NUM];
+	uint32_t single_ended_channels;
```
`uint8_t` should be sufficient (`SAADC_CH_NUM` is 8 in all cases).

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -432,11 +445,42 @@ static int check_buffer_size(const struct adc_sequence *sequence,
+{
+	int8_t channel_id = -1;
+	uint32_t selected_channels = sequence->channels;
+	int16_t *sample_it = nrf_saadc_buffer_pointer_get(NRF_SAADC);
+	int16_t *sample_it_end = sample_it + active_channels;
+
+	do {
+		do {
+			channel_id = channel_id < (SAADC_CH_NUM - 1) ? channel_id + 1 : 0;
+		} while (!(selected_channels & BIT(channel_id)));
+
+		if (!(m_data.single_ended_channels & BIT(channel_id))) {
+			continue;
+		}
+
+		if (*sample_it >= 0) {
+			continue;
+		}
+
+		*sample_it = 0;
+	} while (++sample_it != sample_it_end);
```
Couldn't this be simplified a bit? How about something like this:
```
	uint32_t channel_bit = BIT(0);
	int16_t *sample = nrf_saadc_buffer_pointer_get(NRF_SAADC);

	while (channel_bit < selected_channels) {
		if (selected_channels & channel_bit) {
			if ((single_ended_channels & channel_bit) && *sample < 0) {
				*sample = 0;
			}

			++sample;
		}

		channel_bit <<= 1;
	}
```
You wouldn't need then to pass `active_channels` to this function.

bjarki-andreasen replied:
Right, it was quite complex as I was iterating over every sampling in the sequence, I will clean it up :)

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -96,8 +97,6 @@ static struct driver_data m_data = {
 #if defined(ADC_BUFFER_IN_RAM)
 	.samples_buffer = adc_samples_buffer,
-	.user_buffer = NULL,
-	.active_channels = 0,
```
This change should be actually done in a separate commit as it is not related to negative sample values.

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -525,9 +586,7 @@ static int start_read(const struct device *dev,
 
 	adc_context_start_read(&m_data.ctx, sequence);
 
-	error = adc_context_wait_for_completion(&m_data.ctx);
-
-	return error;
+	return adc_context_wait_for_completion(&m_data.ctx);
```
This one as well.

bjarki-andreasen-nordic committed:
drivers: adc: adc_nrfx_saadc: remove redundant code

Remove redundant copy of error code and 0 initialization of static
m_data struct.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

anangl commented on drivers/adc/adc_nrfx_saadc.c:
```diff
@@ -432,11 +439,37 @@ static int check_buffer_size(const struct adc_sequence *sequence,
+
+static void correct_single_ended(const struct adc_sequence *sequence)
+{
+	uint16_t channel_id = BIT(0);
```
This name may be a bit confusing as the value here is something different from `.channel_id` in `struct adc_channel_cfg` and `channel_id` used in `start_read()`.

bjarki-andreasen replied:
I see, did think the name was odd, I just reused it because of `BIT(channel_id)` without realizing the `BIT()` part of it :D

bjarki-andreasen commented:
ping @kl-cruz :)

aescolar commented:
@jaz1-nordic  @kl-cruz  @magp-nordic   @masz-nordic  @nika-nordic Please take a look
